### ISSUE TITLE
New video with toolbar

### ISSUE THREAD

**Issue by Elv13** [banji-project/ring-issues#6]
Track the requirements and progress. This feature has lower priority than the wizard, but has nevertheless been started and progress between reviews of the wizard. I will now switch back to the wizard as there is some things to change. Here's the current status:

#### Basic features

- [X] Display some video frames in sequences
    - [X] Cache the previous frame in case some are missing (network congestion or buffer bloat)
    - [X] Send black frames when there is too many errors
    - [X] Resize to fit the screen
- [X] Have a toolbar overlay
    - [X] Display contextual actions
    - [X] Is updated on state changes
    - [X] Execute actions when clicked
- [X] Support being a detached popup
- [X] Hide the overlay after 3 seconds when there is no input device activity
- [X] Support custom aspect ratio (black border above and below)
- [X] Display when a call becomes "in progress", hide when it's not
- [X] Drop files into the video for streaming (required to prevent a feature regression)
- [X] Correctly stop the QML acquisition loop
- [X] Find out why the preview remains open after hangup
- [X] Fix the reliability issue as the when the popup is visible
- [X] <s>Find out if the hold feature is broken in the daemon or Ring-KDE</s> (all clients are affected)
- [X] Use the call UserActionModel instead of the global one
- [X] Stop showing the state toolbar in the preview setting page

As shown above, the "big ticket items" are done, but some smaller sub-features still need to be completed before a working prototype can land in the code. The "big risk" items are done, everything else is low risk and requires no additional research/knowledge. 

#### Common / useful features

- [X] Display/toggle the preview
- [ ] Move the preview
- [ ] Rotate /  zoom / pan / freeze
- [ ] Screenshot
    - [ ] Confirmation message
- [ ] Mute/Unmute the video correctly
- [X] Remove the old code
    - [X] Port the settings page to the new code
    - [ ] Port the selfie snapshot for the profile to the new code (the new wizard doesn't have it)

None of those things are priorities, but users will expect them. None of them take much time to implement. Feel free to edit this message and re-order them.

#### Unknowns

 * How to integrate text messaging (in the video canvas or in its own dock/popup)
 * Much more

#### Daemon issues found while testing

After extensive manual testing, the following features seems to fail due to daemon bugs.

 * Sometime the rendering flickers and never comes back
 * Streaming a movie stops the incoming rendering instead of the preview...
 * Unhold and unmute are almost never working
 * Live resolution change doesn't seem correct
 * Some streamed files are not recognized due to their name
 * For some reason if the macOS client calls the KDE/Gnome client there is no media (mac client bug?)

#### Technicalities

I prototyped 2 versions. The first one use an [OpenGL underlay](http://doc.qt.io/qt-5/qtquick-scenegraph-openglunderqml-example.html). It doesn't seem to work correctly for my framebuffer. For some reason it only show a black background. The second uses a [QQuickImageProvider](http://doc.qt.io/qt-5/qquickimageprovider.html#requestTexture). It works fine, but has extra overhead, takes more CPU and currently has a fixed frame rate. I will go ahead with the second prototype for now. I don't want to waste time on OpenGL issues while the other basic features are not ready. A QVideoFrame/QAbstractVideoBuffer version should also be researched / prototyped (eventually). Early versions of QtMultimedia were not suitable to display the GNU Ring / SFLPhone video feed. This may have changed. It comes down to the precise "dance" and mutex synchronization between multiple process. I need total and absolute control over the memory lifetime or the `uchar[]` will be auto-discarded while Qt is reading it, causing a segfault.

#### Other requirements

> Whats definitely good there is the small left sidebar with options, of
> which we only need:
> - last calls (including new call option)
> - adressbook/contacts (including add/remove/edit contacts)
> - settings

This already exists in the other docks and can be combined with the video. The KDE mobile client has the QML version of those lists docks. I can enable them in the video canvas itself if you wish.

[[ADD OTHER REQUIREMENTS HERE]]

**Comment by Elv13** [comment]
Hello,

I pushed the first version of the new video widget. It is an early preview and has the issues listed above.

@shadeslayer Please note Ring-KDE now depends on KF5::Declarative.
@star-buck I am not sure it's good enough for extensive testing yet, but it's working often enough. The best way to see what it looks like is the video preview in settings -> video. The call video itself is still affected by a race condition when it shows up then disappear a second later about 50% of the time. That bug may or may not be in the daemon. I will fix that tomorrow.

**Comment by Elv13** [comment]
The base features are complete and seem mostly reliable on my side.

![pick11](https://cloud.githubusercontent.com/assets/340384/25218572/479a0bb8-2579-11e7-9cdf-0ccc998aca35.png)

**Comment by star-buck** [comment]
this looks awesome, but for some reason, i am stuck when trying to register a global account. will open another ticket and continue comment here when able to get possible issue on my end solved.

**Comment by star-buck** [comment]
i was able to make a call and confirm the popup is showing.
